s3transfer/manager.py

```diff
@@ -378,11 +389,114 @@ def _register_handlers(self):
         self._client.meta.events.register_last(
             event_name, enable_upload_callbacks, unique_id=enable_id)
 
-    def shutdown(self):
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, *args):
+        cancel = False
+        # If a exception was raised in the context handler, signal to cancel
+        # all of the inprogress futures in the shutdown.
+        if exc_type:
+            cancel = True
+        self.shutdown(cancel)
+
+    def shutdown(self, cancel=False):
         """Shutdown the TransferManager
 
-        It will wait till all requests complete before it complete shuts down.
+        It will wait till all transfers complete before it completely shuts
+        down.
+
+        :type cancel: boolean
+        :param cancel: If True, calls TransferFuture.cancel() for
+            all in-progress in transfers. This is useful if you want the
+            shutdown to happen quicker.
+        """
+        if cancel:
+            # Cancel all in-flight transfers if requested, before waiting
+            # for them to complete.
+            self._canceler.cancel()
+        try:
+            # Wait until there are no more in-progress transfers. This is
+            # wrapped in a try statement because this can be interrupted
+            # with a KeyboardInterrupt that needs to be caught.
+            self._canceler.wait()
+        finally:
+            # If not errors were raised in the try block, the cancel should
+            # have no coordinators it needs to run cancel on. If there was
+            # an error raised in the try statement we want to cancel all of
+            # the inflight transfers before shutting down to speed that
+            # process up.
+            self._canceler.cancel()
+            # Shutdown all of the executors.
+            self._submission_executor.shutdown()
+            self._request_executor.shutdown()
+            self._io_executor.shutdown()
+
+
+class TransferCoordinatorCanceler(object):
```

Comment:
Here are other names I can think of:

* TransferCoordinatorManager
* TransferCoordinatorStatusManager
* TransferCoordinatorTracker

Any ideas on these suggestions? I can am leaning toward TransferCoordinatorManager. The other permutation is a bit specific and this class could possibly grow as it can access all public methods of the transfer coordinator.